shared/tests/extension/moduletest/TestModuleTest.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include <catch.hpp>
+#include <nuclear>
+#include <utility>
+
+#include "extension/ModuleTest/ModuleTest.hpp"
+
+// Anon namespace so that this reactor is local to this file
+namespace {
+
+    // We shouldn't emit anything on shutdown, so we'll use this to track whether the shutdown went to plan
+    // This serves as a proxy for testing real on<Shutdown> effects
+    static bool is_shutdown = false;
+
+    // Usually you would use ModuleTest to test an existing reactor with existing messages, but testing ModuleTest
+    // itself requires a reactor and a message, so we define them here
+    template <int id>
+    struct SimpleMessage {
+        SimpleMessage() = default;
+        explicit SimpleMessage(const int& data_) : data(data_) {}
+        int data = 0;
+    };
+
+    // The reactor which is taking the place of the module being tested. When testing a real module, the real module
+    // would take the place of this reactor
+    class TestReactor : public NUClear::Reactor {
+    public:
+        explicit TestReactor(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+            on<Startup>().then([this] {
+                is_shutdown = false;
+                emit(std::make_unique<SimpleMessage<0>>(42));
+            });
+            on<Trigger<SimpleMessage<1>>>().then([this](SimpleMessage<1> /*msg*/) {
+                emit(std::make_unique<SimpleMessage<2>>(42));
+                emit(std::make_unique<SimpleMessage<3>>(31415));
+            });
+            on<Shutdown>().then([this] { is_shutdown = true; });
+        }
+    };
+
+}  // namespace
+
+TEST_CASE("Testing the module tester on the Startup of the TestReactor", "[extension][moduletest][ModuleTest]") {
+
+    // We're testing on<Startup> for this reactor (actually for the ModuleTest), so we disable automatic startup
+    static constexpr bool STARTUP_REACTOR_AUTOMATICALLY = false;
+    auto module_test = extension::moduletest::ModuleTest<TestReactor>(STARTUP_REACTOR_AUTOMATICALLY);
```

Comment:
You have to `install<Reactor>` reactors.

Reply:
They're installed inside the ModuleTest constructor. That's what the template argument is for 😝